File: ruby/lib/resp_bench/config/driver_config.rb
# frozen_string_literal: true

module RespBench
  module Config
    # Configuration for a benchmark driver (client library).
    # Maps to the JSON schema in configs/schemas/driver-config.schema.json
    class DriverConfig
      attr_accessor :schema_version, :description, :driver_id, :mode,
                    :tls, :auth, :specific_driver_config

      def initialize(
        schema_version: "1.0",
        description: nil,
        driver_id: nil,
        mode: "standalone",
        tls: nil,
        auth: nil,
        specific_driver_config: {}
      )
        @schema_version = schema_version
        @description = description
        @driver_id = driver_id
        @mode = mode
        @tls = tls
        @auth = auth
        @specific_driver_config = specific_driver_config || {}
      end

      # Get secondary driver ID for composite drivers (e.g., spring-data-*)
      def secondary_driver_id
        @specific_driver_config["secondary_driver_id"]
      end

      def standalone?
        @mode == "standalone"
      end

      def cluster?
        @mode == "cluster"
      end

      def sentinel?
        @mode == "sentinel"
      end

      def to_h
        {
          schema_version: @schema_version,
          description: @description,
          driver_id: @driver_id,
          mode: @mode,
          tls: @tls,
          auth: @auth,
          specific_driver_config: @specific_driver_config
        }
      end
    end
  end
end

File: python/src/resp_bench/config/driver_config.py
"""Driver (client library) configuration.

Maps to configs/schemas/driver-config.schema.json. Field names and defaults
mirror the Ruby/Java engines so the same JSON files work across all engines.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class DriverConfig:
    schema_version: str = "1.0"
    description: Optional[str] = None
    driver_id: Optional[str] = None
    mode: str = "standalone"
    command_timeout_ms: Optional[int] = None
    tls: Optional[dict] = None
    auth: Optional[dict] = None
    specific_driver_config: dict = field(default_factory=dict)

    def secondary_driver_id(self) -> Optional[Any]:
        """Secondary driver id for composite drivers (e.g. spring-data-*)."""
        return self.specific_driver_config.get("secondary_driver_id")

    def is_standalone(self) -> bool:
        return self.mode == "standalone"

    def is_cluster(self) -> bool:
        return self.mode == "cluster"

    def is_sentinel(self) -> bool:
        return self.mode == "sentinel"

    def tls_enabled(self) -> bool:
        return bool(self.tls and self.tls.get("enabled"))
